syntax = "proto3";
package IotDeviceManagement2;

service InitialConfiguration {
  rpc RegisterDevice(RegisterDeviceRequest) returns (RegisterDeviceResponse);
  rpc UpdateOwnDevice(UpdateOwnDeviceRequest) returns (UpdateOwnDeviceResponse);
  rpc GetDeviceIdByDeviceName(GetDeviceIdByDeviceNameRequest) returns (GetDeviceIdByDeviceNameResponse);
  rpc ConfigureNetwork (ConfigureNetworkRequest) returns (ConfigureNetworkResponse);
}
service SystemStatusService {
    rpc GetSystemStatus (SystemStatusRequest) returns (SystemStatusResponse);
    rpc GetLastRecord(GetLastRecordRequest) returns (GetLastRecordResponse);
}
service FirmwareUpdateService {
 rpc ApplyFirmwareUpdate (ApplyFirmwareUpdateRequest) returns (ApplyFirmwareUpdateRequest);
}



message RegisterDeviceRequest {
  string serial_number = 1;
  string name = 2;
  string type = 3;
  string location = 4;
  string owner = 5;
  string os_type = 6;
}

message RegisterDeviceResponse {
  string message = 1;
  int32 device_id = 2;
}

message UpdateOwnDeviceRequest {
  int32 device_id = 1;
  string name = 2;
  string owner = 3;
  string os_type = 4;
  string location = 5;
}

message UpdateOwnDeviceResponse {
  string message = 1;
}

message GetDeviceIdByDeviceNameRequest {
  string device_name = 1;
}

message GetDeviceIdByDeviceNameResponse {
  int32 device_id = 1;
}


// Messages for ConfigureNetwork
message ConfigureNetworkRequest {
  int32 device_id = 1;
  string ssid = 2;
  string wifi_password = 3;
  string ip_address = 4;
}

message ConfigureNetworkResponse {
  string message = 1;
}



message SystemStatusRequest {
    int32 device_id = 1;
    string cpu_usage = 2;
    string memory_usage = 3;
    string disk_space = 4;
}

message SystemStatusResponse {

    string message = 1;
}
message GetLastRecordRequest {
    int32 device_id = 1;
}

message GetLastRecordResponse {
    string cpu_usage = 1;
    string memory_usage = 2;
    string disk_space = 3;
    string timestamp = 4;
    string message = 5;
}
message ApplyFirmwareUpdateRequest {
    int32 device_id = 1;
}

message ApplyFirmwareUpdateResponse {
    bytes firmware_data = 1;
    string message = 2;
}



